### DISCUSSION TITLE
Improve the usability of pullback parameters of skrf.calibration.deembedding.IEEEP370_MM_ZC_2xThru

### DISCUSSION THREAD

**Discussion by z331225718** [scikit-rf/scikit-rf#1075] [Ideas]
Now pullback represents a number of discrete points to leave in the fixture. In the limited bandwidth, the length interval is too large.
Is it possible to improve the accuracy and turn this parameter into a more accurate length?

**Comment by mhuser** [comment]
Hello @z331225718 

The [EEEP370_SE_ZC_2xThru](https://scikit-rf.readthedocs.io/en/latest/api/calibration/generated/skrf.calibration.deembedding.IEEEP370_SE_ZC_2xThru.html) or its mixed-modes generalization works by peeling the half-fixture point by point and model each layer by a length of transmission line of a suitable impedance. This operation is done in a loop were the chunks of transmission lines are deembedded iteratively from the fixture, which effectively "peel" it.

This is the reason why pullback is currently a discrete number, with an unity length corresponding to one transmission line chunk length, which has the length of one sample in time-domain.

https://github.com/scikit-rf/scikit-rf/blob/8461e77f61d1bfae595dacc2544a16c3c25453ee/skrf/calibration/deembedding.py#L2072-L2088

I am not sure I have properly understood the last part of your idea:

> In the limited bandwidth, the length interval is too large.
Is it possible to improve the accuracy and turn this parameter into a more accurate length?

**Reply by mhuser** [reply]
It is maybe possible to peel a last chunk of line of a fraction of the sample length if the pullback has a fractionnary part 🤔 
But I have no dataset to test such a wild thing.

**Reply by z331225718** [reply]
I haven't looked at the implementation process carefully. At a rough glance, I don't quite understand why a loop is used instead of calculating the total length at once and then directly using makeTL(zline1,z0,gamma,l_total). In this way,we can customize any length?

**Reply by z331225718** [reply]
I modify the code:
```
def makeErrorBox_v7_2(self, s_dut, s2x, gamma, z0, pullback):
        f = s2x.frequency.f
        n = len(f)
        s212x = s2x.s[:, 1, 0]
        DC21 = self.dc_interp(s212x, f)
        x = np.argmax(irfft(concatenate(([DC21], s212x))))
        #define relative length
        #python first index is 0, thus 1 should be added to get the length
        #l = 1. / (2 * (x + 1))
        l_fixture = 1-pullback
        #define the reflections to be mimicked
        s11dut = s_dut.s[:, 0, 0]
        s22dut = s_dut.s[:, 1, 1]
        #peel the fixture away and create the fixture model
        #python range to n-1, thus 1 to be added to have proper iteration number
        zline1 = self.getz(s11dut, f, z0)[0]
        zline2 = self.getz(s22dut, f, z0)[0]
        TL1 = self.makeTL(zline1,z0,gamma,l_fixture)
        TL2 = self.makeTL(zline2,z0,gamma,l_fixture)
        sTL1 = s_dut.copy()
        sTL1.s = TL1
        sTL2 = s_dut.copy()
        sTL2.s = TL2
        errorbox1 = errorbox1 ** sTL1
        errorbox2 = errorbox2 ** sTL2
            # equivalent to function removeTL(in,TL1,TL2,z0)
            # no need to flip sTL2 because it is symmetrical
        s_dut = sTL1.inv ** s_dut ** sTL2.inv
            #IEEE abcd implementation
            # abcd_TL1 = sTL1.a
            # abcd_TL2 = sTL2.a
            # abcd_in  = s_dut.a
            # for j in range(len(s_dut.frequency.f)):
            #     abcd_in[j, :, :] = np.linalg.lstsq(abcd_TL1[j, :, :].T, np.linalg.lstsq(abcd_TL1[j, :, :], abcd_in[j, :, :], rcond=None)[0].T, rcond=None)[0].T
            # s_dut.a = abcd_in
        s11dut = s_dut.s[:, 0, 0]
        s22dut = s_dut.s[:, 1, 1]
```
tune the input pullback can get more accurate result

**Comment by denzchoe** [comment]
Hello, I want to sidetrack a little bit.
The actual implementation of the code was inherited from the IEEE P370 Standard's Repo.
You can see the link below of the original code..
https://gitlab.com/IEEE-SA/ElecChar/P370/-/blob/master/TG1/IEEEP370Zc2xThru.m#L347 

and then the method was introduced from this PR
https://gitlab.com/IEEE-SA/ElecChar/P370/-/merge_requests/11?commit_id=6a3aff5562dccc75b6d64b5747dbdf47c8eb865a

**Comment by mhuser** [comment]
There is a loop because the fixture is modeled by a serie of short transmission lines of **different** impedances. The loop peel the fixture to mimic the time-domain response impedance by impedance.

Without the loop, it looks more like a kind of port extension, with a single transmission line. I think it is not the purpose of the method.

Indeed it can gives good results in some particuliar cases and the method provides the base parameters of the transmission line. But it is no more IEEE P370 as @denzchoe pointed out.

**Reply by z331225718** [reply]
If for some reason, we need to manually add a uniform impedance extension line between the 2xthru fixtures, can we modify the original method and use it? This is also the main reason why I asked this question. In fact, it is a common scenario in multi-channel highspeed signal test.

**Reply by mhuser** [reply]
If I understand properly, this means to extend both fixtures on DUT side (port 2 in schematic below) with an uniform transmission line of a given length and the last impedance of the fixture, after the 2xthru was split into two halves, resulting in longer fixtures than half 2xthru?

The pullback lead instead to a shorter fixture (it leave some of the fixture in the dembedded results), so it probably does the opposite. Thus I was at first confused.

```
         FIX-1    DUT      FIX-2
         +----+   +----+   +----+
        -|1  2|---|1  2|---|2  1|-
         +----+   +----+   +----+
```

**Reply by z331225718** [reply]
I think there are two usage scenarios. One is as shown in the picture. It is necessary to embed fixed connectors and several sections of transmission lines, but there is no way to accurately provide 2xthru in a structure. There must be some signals that need to be extended, but this part of the extension line If added to the fixture, it will cause errors. 
<img width="1321" alt="企业微信截图_17169518034013" src="https://github.com/scikit-rf/scikit-rf/assets/12806611/382701b6-31cd-44c0-a77f-62e3e0b59931">
The other is if there is an impedance transformation near the middle area of ​​2xthru. For example, some chips have a small fan-out area and require thin wires. In this scenario, if the length of the thin wire is not within dl, errors may occur.
I modify the code:
```
for i in range(x):#shortern by 1
            zline1 = self.getz(s11dut, f, z0)[0]
            zline2 = self.getz(s22dut, f, z0)[0]
            TL1 = self.makeTL(zline1,z0,gamma,l)
            TL2 = self.makeTL(zline2,z0,gamma,l)
            sTL1 = s_dut.copy()
            sTL1.s = TL1
            sTL2 = s_dut.copy()
            sTL2.s = TL2
            if i == 0:
                errorbox1 = sTL1
                errorbox2 = sTL2
                _, z1 = self.getz(s11dut, f, z0)
                _, z2 = self.getz(s22dut, f, z0)
            else:
                errorbox1 = errorbox1 ** sTL1
                errorbox2 = errorbox2 ** sTL2
            # equivalent to function removeTL(in,TL1,TL2,z0)
            # no need to flip sTL2 because it is symmetrical
            s_dut = sTL1.inv ** s_dut ** sTL2.inv
            #s2x = sTL1.inv ** s2x ** sTL2.inv
            #gamma = self.get_gamma(s2x)
            #IEEE abcd implementation
            # abcd_TL1 = sTL1.a
            # abcd_TL2 = sTL2.a
            # abcd_in  = s_dut.a
            # for j in range(len(s_dut.frequency.f)):
            #     abcd_in[j, :, :] = np.linalg.lstsq(abcd_TL1[j, :, :].T, np.linalg.lstsq(abcd_TL1[j, :, :], abcd_in[j, :, :], rcond=None)[0].T, rcond=None)[0].T
            # s_dut.a = abcd_in
            s11dut = s_dut.s[:, 0, 0]
            s22dut = s_dut.s[:, 1, 1]
            if self.verbose:
                if i == 0:
                    fig, axs = subplots(2, 2)
                    axs[0, 0].plot(z1, color = 'k')
                    axs[0, 0].set_xlim((0, x))
                    axs[0, 1].plot(z2, color = 'k')
                    axs[0, 1].set_xlim((0, x))
                    axs[1, 0].set_xlim((n-100, n+x*2+10))
                    axs[1, 1].set_xlim((n-100, n+x*2+10))
                _, zeb1 = self.getz(errorbox1.s[:, 0, 0], f, z0)
                _, zeb2 = self.getz(errorbox2.s[:, 0, 0], f, z0)
                _, zdut1 = self.getz(s11dut, f, z0)
                _, zdut2 = self.getz(s22dut, f, z0)
                axs[0, 0].plot(zeb1)
                axs[0, 1].plot(zeb2)
                axs[1, 0].plot(ifftshift(zdut1))
                axs[1, 1].plot(ifftshift(zdut2))

        l_pullback1 = pullback-l
        l_pullback2 = pullback-l
        zline_pullback1 = self.getz(s11dut, f, z0)[0]
        zline_pullback2 = self.getz(s22dut, f, z0)[0]
        TL_pullback1 = self.makeTL(zline_pullback1,z0,gamma,l_pullback1)
        TL_pullback2 = self.makeTL(zline_pullback2,z0,gamma,l_pullback2)
        sTL_pullback1 = s_dut.copy()
        sTL_pullback2 = s_dut.copy()
        sTL_pullback1.s = TL_pullback1
        sTL_pullback2.s = TL_pullback2
        errorbox1 = errorbox1 ** sTL_pullback1.inv
        errorbox2 = errorbox2 ** sTL_pullback2.inv
```
It's useful in my test snps(pullback value may be calculated by 2 snps' group delay ratio, tuned by users)
<img width="400" alt="企业微信截图_1716949221664" src="https://github.com/scikit-rf/scikit-rf/assets/12806611/dbae7806-e7ef-48a5-923e-be57aa76d723">

**Reply by denzchoe** [reply]
@z331225718 , `dm_fixture1` is simulated or deembedded from measurement ?

**Reply by z331225718** [reply]
@denzchoe de-embeded from simulated result(2xthru with a extension line) to verify pullback

**Reply by mhuser** [reply]
Ok, I think I got it. Thanks for the explanations.

Can you try the code below? The pullback is still given in discrete sample units to keep backward compatibility but it may be a fractional number. Thus it keep shortening the peeling loop. The last transmission line slice, however, has its length finetuned as a fraction of the base length based on the fractional part of pullback.

```Python
pullback1 = 5.5, pullback2 = 5.5
or
pullback1 = 0.12, pullback2 = 0.12
and so on
```

```Python
from math import modf
def makeErrorBox_v7(self, s_dut, s2x, gamma, z0, pullback):
        f = s2x.frequency.f
        n = len(f)
        # fine and coarse pullback
        pullback_frac, pullback_int = modf(pullback)
        pullback_int = int(pullback_int)
        s212x = s2x.s[:, 1, 0]
        DC21 = self.dc_interp(s212x, f)
        x = np.argmax(irfft(concatenate(([DC21], s212x))))
        #define relative length
        #python first index is 0, thus 1 should be added to get the length
        l = 1. / (2 * (x + 1))
        #define the reflections to be mimicked
        s11dut = s_dut.s[:, 0, 0]
        s22dut = s_dut.s[:, 1, 1]
        #peel the fixture away and create the fixture model
        #python range to n-1, thus 1 to be added to have proper iteration number
        # 1 is removed to enable fine fractionnal pull-back after the loop
        for i in range(x - pullback_int):
            zline1 = self.getz(s11dut, f, z0)[0]
            zline2 = self.getz(s22dut, f, z0)[0]
            TL1 = self.makeTL(zline1,z0,gamma,l)
            TL2 = self.makeTL(zline2,z0,gamma,l)
            sTL1 = s_dut.copy()
            sTL1.s = TL1
            sTL2 = s_dut.copy()
            sTL2.s = TL2
            if i == 0:
                errorbox1 = sTL1
                errorbox2 = sTL2
                _, z1 = self.getz(s11dut, f, z0)
                _, z2 = self.getz(s22dut, f, z0)
            else:
                errorbox1 = errorbox1 ** sTL1
                errorbox2 = errorbox2 ** sTL2
            # equivalent to function removeTL(in,TL1,TL2,z0)
            # no need to flip sTL2 because it is symmetrical
            s_dut = sTL1.inv ** s_dut ** sTL2.inv
            #IEEE abcd implementation
            # abcd_TL1 = sTL1.a
            # abcd_TL2 = sTL2.a
            # abcd_in  = s_dut.a
            # for j in range(len(s_dut.frequency.f)):
            #     abcd_in[j, :, :] = np.linalg.lstsq(abcd_TL1[j, :, :].T,
            #                                        np.linalg.lstsq(abcd_TL1[j, :, :], abcd_in[j, :, :],
            #                                        rcond=None)[0].T, rcond=None)[0].T
            # s_dut.a = abcd_in
            s11dut = s_dut.s[:, 0, 0]
            s22dut = s_dut.s[:, 1, 1]
            if self.verbose:
                if i == 0:
                    fig, axs = subplots(2, 2)
                    axs[0, 0].plot(z1, color = 'k')
                    axs[0, 0].set_xlim((0, x))
                    axs[0, 1].plot(z2, color = 'k')
                    axs[0, 1].set_xlim((0, x))
                    axs[1, 0].set_xlim((n-100, n+x*2+10))
                    axs[1, 1].set_xlim((n-100, n+x*2+10))
                _, zeb1 = self.getz(errorbox1.s[:, 0, 0], f, z0)
                _, zeb2 = self.getz(errorbox2.s[:, 0, 0], f, z0)
                _, zdut1 = self.getz(s11dut, f, z0)
                _, zdut2 = self.getz(s22dut, f, z0)
                axs[0, 0].plot(zeb1)
                axs[0, 1].plot(zeb2)
                axs[1, 0].plot(ifftshift(zdut1))
                axs[1, 1].plot(ifftshift(zdut2))
        # last step with fine pull-back
        zline1 = self.getz(s11dut, f, z0)[0]
        zline2 = self.getz(s22dut, f, z0)[0]
        TL1 = self.makeTL(zline1,z0,gamma,(1. - pullback_frac) * l)
        TL2 = self.makeTL(zline2,z0,gamma,(1. - pullback_frac) * l)
        sTL1 = s_dut.copy()
        sTL1.s = TL1
        sTL2 = s_dut.copy()
        sTL2.s = TL2
        errorbox1 = errorbox1 ** sTL1
        errorbox2 = errorbox2 ** sTL2
        return errorbox1, errorbox2.flipped()
```

To modifiy case were the pullbacks are not equals, `makeErrorBox_v8` would also need to be reworked.
